packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+ * bindings.
+ */
+type CustomElementClosedOp = {
+  type: 'custom-element-close';
+};
+
+type Op =
+  | TextOp
+  | NodePartOp
+  | AttributePartOp
+  | CustomElementOpenOp
+  | CustomElementAttributesOp
+  | CustomElementChildrenOp
+  | CustomElementClosedOp;
+
+const getTemplate = (result: TemplateResult) => {
+  const template = templateCache.get(result.strings);
+  if (template !== undefined) {
+    return template;
+  }
+  const {html, attrNames} = getTemplateHtml(result.strings, result._$litType$);
+  const ast = parseFragment(html, {
+    sourceCodeLocationInfo: true,
+  }) as DefaultTreeDocumentFragment;
+
+  const ops: Array<Op> = [];
+
+  /* The last offset of html written to the stream */
+  let lastOffset: number | undefined = 0;
+
+  /**
+   * Sets `lastOffset` to `offset`, skipping a range of characters. This is
+   * useful for skipping <slot>s and distributed nodes in flattened mode, or
+   * skipping and re-writting lit-html marker nodes.
+   */
+  const skipTo = (offset: number) => {
+    if (lastOffset === undefined) {
+      throw new Error('lastOffset is undefined');
+    }
+    if (offset < lastOffset) {
+      throw new Error(`offset must be greater than lastOffset.
+        offset: ${offset}
+        lastOffset: ${lastOffset}
+      `);
+    }
+    lastOffset = offset;
+  };
+
+  /**
+   * Creates or appends to a text opcode with a substring of the html from the
+   * `lastOffset` flushed to `offset`.
+   */
+  const flushTo = (offset?: number) => {
+    if (lastOffset === undefined) {
+      throw new Error('lastOffset is undefined');
+    }
+    const previousLastOffset = lastOffset;
+    lastOffset = offset;
+    const value = html.substring(previousLastOffset, offset);
+    const op = getLast(ops);
+    if (op !== undefined && op.type === 'text') {
+      op.value += value;
+    } else {
+      ops.push({
+        type: 'text',
+        value,
+      });
+    }
+  };
+
+  const flush = (value: string) => {
+    const op = getLast(ops);
+    if (op !== undefined && op.type === 'text') {
+      op.value += value;
+    } else {
+      ops.push({
+        type: 'text',
+        value,
+      });
+    }
+  };
+
+  // Depth-first node index. Initialized to -1 so that the first child node is
+  // index 0, to match client-side lit-html.
+  let nodeIndex = -1;
+
+  traverse(ast, {
+    pre(node, parent) {
+      if (isCommentNode(node)) {
+        if (node.data === markerMatch) {
+          flushTo(node.sourceCodeLocation!.startOffset);
+          skipTo(node.sourceCodeLocation!.endOffset);
+          ops.push({
+            type: 'node-part',
+            index: nodeIndex,
+            useCustomElementInstance:
+              parent && isElement(parent) && parent.isDefinedCustomElement,
+          });
+          // There will be two comments per part (open+close) in the rendered
+          // output and on the client, so increment again for that
+          // nodeIndex++;
```

Comment:
Need to confirm why this is no longer true, but likely due to the client-side change to only make one marker for node parts.

Reply:
yep